analytics/client.py

```diff
@@ -34,6 +35,12 @@ def __init__(self, write_key=None, debug=False, max_queue_size=10000,
         self.debug = debug
         self.send = send
 
+        # On program exit, allow the consumer thread to exit cleanly.
+        # This prevents exceptions and a messy shutdown when the interpreter is
+        # destroyed before the daemon thread finishes execution. However, it
+        # is *not* the same as flushing the queue! To guarantee all
+        atexit.register(self.join)
```

Comment:
It is almost the behavior that we want, which is why this is still a daemon thread! The reason for this change is that we want to allow the thread to exit cleanly. When we call `self.join` here we set `running` to `False` in the consumer, and it will exit after its current `upload` loop.

Without this change (in the current behavior), the result will (often but not always) be a cryptic exception thrown as the program is exiting. It is hard to debug, too, because the interpreter is also shutting down -- even the stack trace information itself is not preserved.

This is a common issue people run into when using daemon threads; you have to think about how they are going to clean up.
https://github.com/segmentio/analytics-python/issues/46 has a link to a good description of the issue.